File: app/src/main/java/com/syroniko/casseteapp/LogInSignUp/GenrePickSignupAdapter.java
package com.syroniko.casseteapp.LogInSignUp;


import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;
import com.syroniko.casseteapp.R;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class GenrePickSignupAdapter extends RecyclerView.Adapter<GenrePickSignupAdapter.GenreViewHolder> {

    private List<GenreNameImageForSignupAdapter> genreList;
    private GenreListClickListener genreListClickListener;


    GenrePickSignupAdapter(List<GenreNameImageForSignupAdapter> genreList, GenreListClickListener genreListClickListener) {
        this.genreList = genreList;
        this.genreListClickListener=genreListClickListener;
    }

    @NotNull
    @Override
    public GenreViewHolder onCreateViewHolder(@NotNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.genre_signup_item,viewGroup,false );
        return new GenreViewHolder(view);
    }


    @Override
    public void onBindViewHolder(@NotNull GenreViewHolder holder, int position) {
        GenreNameImageForSignupAdapter genre = genreList.get(position);
        holder.textView.setText(genre.getGenre());
        holder.imageView.setImageResource(genre.getGenreImageDefault());
    }

    @Override
    public int getItemCount() {
        return genreList.size();
    }

    public interface GenreListClickListener{
        void OnListItemClick(int clickedItemPosition);
    }

    class GenreViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener{
        private TextView textView;
        private ImageView imageView;

        GenreViewHolder(View view){
            super(view);
            textView = view.findViewById(R.id.genre_string_item);
            imageView = view.findViewById(R.id.genre_image_signup);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            int clickedPosition = getAdapterPosition();

            genreListClickListener.OnListItemClick(clickedPosition);
            genreList.get(clickedPosition).setClicked(!genreList.get(clickedPosition).isClicked());

            Boolean state = genreList.get(clickedPosition).isClicked();

            if (!state) {
                imageView.setImageResource(genreList.get(clickedPosition).getGenreImageDefault());
            }
            else{
                imageView.setImageResource(genreList.get(clickedPosition).getGenreImageClicked());
            }
        }

    }
}

File: app/src/main/java/com/syroniko/casseteapp/logInSignUp/GenrePickSignupAdapter.kt
package com.syroniko.casseteapp.logInSignUp

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.syroniko.casseteapp.logInSignUp.GenrePickSignupAdapter.GenreViewHolder
import com.syroniko.casseteapp.R

class GenrePickSignupAdapter internal constructor(
    private val genreList: List<GenreNameImageForSignupAdapter>,
    private val onClick: (Int) -> Unit
) : RecyclerView.Adapter<GenreViewHolder>() {

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): GenreViewHolder {
        val view = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.genre_signup_item, viewGroup, false)

        return GenreViewHolder(view)
    }

    override fun onBindViewHolder(holder: GenreViewHolder, position: Int) {
        val genre = genreList[position]
        holder.textView.text = genre.genre

        holder.imageView.setImageResource(genre.genreImageDefault)
        holder.imageView.setOnClickListener {
            onImageClick(position, it as ImageView)
        }
    }

    override fun getItemCount() = genreList.size

    private fun onImageClick(position: Int, imageView: ImageView){
        onClick(position)

        genreList[position].isClicked = !genreList[position].isClicked

        val clicked = genreList[position].isClicked
        if (clicked) {
            imageView.setImageResource(genreList[position].genreImageClicked)
        }
        else {
            imageView.setImageResource(genreList[position].genreImageDefault)
        }
    }

    class GenreViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val textView: TextView = view.findViewById(R.id.genre_string_item)
        val imageView: ImageView = view.findViewById(R.id.genre_image_signup)
    }

}